File: C++_Version/String/557Reverse_Words_in_a_String_III.cpp
//Given a string, you need to reverse the order of characters in each word within a sentence while still preserving whitespace and initial word order.
//
//Example 1:
//Input: "Let's take LeetCode contest"
//	Output : "s'teL ekat edoCteeL tsetnoc"
//	Note : In the string, each word is separated by single space and there will not be any extra space in the string.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <stack>
using namespace std;


class Solution {
public:
	string reverseWords(string s) {

		int length = s.length();
		int start = 0;

		for (int i = 0; i < length; ++i)
		{
			if (s[i] == ' ')
			{
				reverse(s.begin() + start, s.begin() + i);
				start = i + 1;
			}
			if (i == length - 1)
				reverse(s.begin() + start, s.end());
		}

		return s;
	}
};

int main()
{
	string string1 = "hello darkness my old friend", string2 = "el";

	Solution sol1;

	string receive_value = sol1.reverseWords(string1);

	cout << receive_value << endl;

	system("pause");

	return 0;
}

File: Java_Version/557Reverse_Words_in_a_String_III.java
/*Given a string, you need to reverse the order of characters in each word within a sentence while still preserving whitespace and initial word order.

        Example 1:
        Input: "Let's take LeetCode contest"
        Output: "s'teL ekat edoCteeL tsetnoc"
        Note: In the string, each word is separated by single space and there will not be any extra space in the string.*/


import java.util.*;
import java.lang.Math;
import java.lang.System;
import java.lang.Integer;


public class Main {

    public static void main(String[] args) throws ArithmeticException{

        int a = 3, b = 0;


        int result = a/b;


        System.out.println("haha");
    }

}


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {
    public String reverseWords(String s) {

        String[] splitS = s.split("\\s");

        int splitSLength = splitS.length;
        if (splitSLength < 1){
            return "";
        }

        StringBuilder result = new StringBuilder();

        for (int i = 0 ; i < splitSLength - 1; i++){
            result = result.append(new StringBuilder(splitS[i]).reverse()).append(' ');
        }

        return result.append(new StringBuilder(splitS[splitSLength-1]).reverse()).toString();
    }
}